lib/std/special/compiler_rt/emutls.zig

```diff
@@ -0,0 +1,356 @@
+        if (self.slots[index] == null) {
+            // initialize the slot
+            const size = control.size;
+            const alignment = std.math.max(
+                @truncate(u29, control.alignment),
+                @alignOf(usize),
+            );
+
+            // XXX slot allocated using posix_memalign
+            var aligned_ptr: ?*c_void = undefined;
+            if (std.c.posix_memalign(&aligned_ptr, alignment, size) != 0) {
+                @panic("posix_memalign");
+            }
+            errdefer std.c.free(aligned_ptr);
+
+            if (control.default_value) |value| {
+                // default value: copy the content to newly allocated object.
+                @memcpy(@ptrCast([*]u8, aligned_ptr), @ptrCast([*]u8, value), size);
+            } else {
+                // no default: return zeroed memory.
+                @memset(@ptrCast([*]u8, aligned_ptr), 0, size);
+            }
+
+            self.slots[index] = @ptrCast(*c_void, aligned_ptr);
+        }
+
+        return self.slots[index].?;
+    }
+};
+
+// global stucture for ThreadStorage.
+// it provides thread-safety for on-demand storage of Thread Objects.
+const ThreadStorage = struct {
+    var key: std.c.pthread_key_t = undefined;
+    var init_once = std.once(ThreadStorage.init);
+
+    /// Return a per thread ObjectArray with at least the expected index.
+    pub inline fn get_array(allocator: *Allocator, index: usize) !*ObjectArray {
+        if (ThreadStorage.getspecific()) |array| {
+            // ensure the array is big enough for the wanted index
+            try array.ensureSize(index);
+            return array;
+        } else {
+            // a new array should contains at least 16 objects.
+            const size = std.math.max(16, index);
+
+            // create a new array and store it.
+            var array: *ObjectArray = try ObjectArray.initAlloc(allocator, size);
+            ThreadStorage.setspecific(array);
+            return array;
+        }
+    }
+
+    /// Return casted thread specific value.
+    inline fn getspecific() ?*ObjectArray {
+        return @ptrCast(
+            ?*ObjectArray,
+            @alignCast(
+                @alignOf(?*ObjectArray),
+                std.c.pthread_getspecific(ThreadStorage.key),
+            ),
+        );
+    }
+
+    /// Set casted thread specific value.
+    inline fn setspecific(new: ?*ObjectArray) void {
+        if (std.c.pthread_setspecific(ThreadStorage.key, @ptrCast(*c_void, new)) != 0)
+            @panic("pthread_setspecific");
+    }
+
+    /// Initialize pthread_key_t.
+    fn init() void {
+        if (std.c.pthread_key_create(&ThreadStorage.key, ThreadStorage.deinit) != 0)
+            @panic("pthread_key_create");
+    }
+
+    /// Invoked by pthread specific destructor. the passed argument is the ObjectArray pointer.
+    fn deinit(arrayPtr: *c_void) callconv(.C) void {
+        var array = @ptrCast(
```

Comment:
I think it's about time this `@ptrCast(*T, @alignCast(@alignOf(T), ptr)` pattern is turned into a nice helper in `std.mem`.

And... are you sure `@alignOf(*ObjectArray)` is not `@alignOf(ObjectArray)` ?

Reply:
> And... are you sure @alignOf(*ObjectArray) is not @alignOf(ObjectArray) ?

the passed argument is the same object set with `pthread_setspecific()`. I passed a `*ObjectArray`, so I should get a `*ObjectArray`. Regarding the alignment, I am a bit unexperimented with them. I am casting to `*ObjectArray` so I keep the same for the alignment. But I could be wrong.

Reply:
This small snippet should help understand the subtle difference between the two:

```zig
    const A = struct {
        x: u32,
    };
    var x: A = undefined;
    @compileLog(@typeInfo(@TypeOf(&x)).Pointer.alignment); // 4
    @compileLog(@typeInfo(@TypeOf(@alignCast(@alignOf(A), &x))).Pointer.alignment); // 4
    @compileLog(@typeInfo(@TypeOf(@alignCast(@alignOf(*A), &x))).Pointer.alignment); // 8
```